### PR TITLE
Fix/socket io design changes

### PR DESCRIPTION
**Pull request by muhammad-ali-e**
## What

- Updated SocketIO logging from pubsub to celery queue
- Implemented socket rooms
- Partial implementation of session authentication in socket connection
- Implemented Connection Manager

## Why

- Addressing drawbacks in the current WebSocket logs:
  - No SessionId check, allowing SocketIO connections even from the landing page before login.
  - All emitted messages from the server are received by all connected clients, causing data leaks.
  - Each message from publisher is received by every backend worker due to pubsub
- The Connection Manager enables using the same socket across different services or sessions. 

## How

- Changed log consumer to Celery tasks.
-Replaced pubsub with Kombu Publisher.
- Implemented SocketIO rooms for each user session to prevent data leaks.
- Added Session ID null check

## Can this PR break any existing features. If yes, please list possible items. If no, please explain why. (PS: Admins do not merge the PR without this section filled)

- No, the main changes are in the log publisher. We need to verify that no logs are missing.
## Database Migrations

-  No

## Env Config

- Backend:
  -  NEW
    -  CELERY_BROKER_URL
- Worker
  - NEW
    - CELERY_BROKER_URL
  - TO REMOVE
    - REDIS_HOST
    - REDIS_PORT
    - REDIS_PASSWORD
    - REDIS_USER

## Relevant Docs

-

## Related Issues or PRs

-

## Dependencies Versions

-

## Notes on Testing

-

## Screenshots
Screenshot from 2024-06-05 14-19-26

## Checklist

I have read and understood the [Contribution Guidelines]().

### CODE REVIEW

File: backend/backend/celery_task.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+logger = logging.getLogger(__name__)
+
+
+class TaskRegistry:
+    @staticmethod
+    @shared_task(
+        name=LogProcessingTask.TASK_NAME,
+    )
+    def log_consumer(**kwargs: Any) -> None:
```
Comment (chandrasekharan-zipstack) [inline backend/backend/celery_task.py:18]: NIT: Please add a docstring highlighting the kwargs

Comment (muhammad-ali-e) [reply backend/backend/celery_task.py:18]: @chandrasekharan-zipstack Updated. I was missed entire docstring

**Review by jaseemjaskp — APPROVED** [review]

File: unstract/core/src/unstract/core/pubsub_helper.py [hunk 2]
```diff
@@ -93,12 +90,53 @@ def log_prompt(
...
+                # Publish the message to the queue
+                producer.publish(
+                    body=task_message,
+                    exchange="",
+                    headers=headers,
+                    routing_key=LogProcessingTask.QUEUE_NAME,
+                    compression=None,
+                    retry=True,
+                )
+                logging.info(f"Published '{channel_id}' <= {payload}")
```
Comment (chandrasekharan-zipstack) [inline unstract/core/src/unstract/core/pubsub_helper.py:138]: NIT: This could be a debug log if its too noisy

**Review by chandrasekharan-zipstack — APPROVED** [review]
LGTM - left some nit comments

File: backend/backend/celery_service.py [hunk 3]
```diff
@@ -15,10 +17,14 @@
 # Create a Celery instance. Default time zone is UTC.
 app = Celery("backend")
 
+# To register the custom tasks.
+TaskRegistry()
```
[the remark this answers is not in the document]
Comment (muhammad-ali-e) [reply backend/backend/celery_service.py:21]: The shared task will not pickup if it not a part of django backend or part of this file. So it have to initiate here.

File: backend/pyproject.toml [hunk 4]
```diff
@@ -78,7 +78,7 @@ migrate_db.cmd = "python manage.py migrate"
 migrate_db.env_file = ".env"
 migrate_db.help = "Performs DB migrations for Unstract backend"
 # Commands for backend execution
-consumer.cmd = "celery -A backend worker --loglevel=info -Q celery,celery_periodic_logs"
+consumer.cmd = "celery -A backend worker --loglevel=info -Q celery,celery_periodic_logs,celery_log_task_queue --autoscale 8,1"
```
[the remark this answers is not in the document]
Comment (muhammad-ali-e) [reply backend/pyproject.toml:81]: Already have

File: unstract/core/src/unstract/core/pubsub_helper.py (hunk 2 above)
Comment (muhammad-ali-e) [reply unstract/core/src/unstract/core/pubsub_helper.py]: updated

File: backend/utils/log_events.py [hunk 5]
```diff
@@ -109,69 +144,36 @@ def _store_execution_log(data: bytes) -> None:
...
+        room (str): Room to emit event to
+        event (str): Event name
+        channel (str): Channel name
+        data (bytes): Data to emit
     """
     payload = {"data": data}
     try:
-        logger.debug(f"[{os.getpid()}] Push websocket event: {channel}, {payload}")
-        sio.emit(channel, payload)
+        logger.debug(f"[{os.getpid()}] Push websocket event: {event}, {payload}")
```
[the remark this answers is not in the document]
Comment (muhammad-ali-e) [reply backend/utils/log_events.py:157]: We can check it later once we check the logs

**Pull request merged**